linode/manager
# feat: [M3-8722] - Improve weblish retry behavior

## Description 📝

- automatically reconnect weblish sessions that have disconnected
- … unless it fails rapidly (more than 3 times in a 1 minute window)
- … in which case show an error screen with as much information as we can provide
- … where the error screen includes a button to try again.
- focus the terminal element when starting (which is generally helpful but also prevents _losing_ focus after a reconnect)

Previously, a failure would often just result in a spinning circle forever.

## Changes  🔄
List any change relevant to the reviewer.
- added option “action button” feature to ErrorState
- added some helpers to `Lish.tsx` which may also be useful for similar enhancements to graphical lish (glish) in the future.
- changed `socket` to be mutable and nullable so it can be updated when reconnecting. `origSocket` local variable is used to ensure that “stale” events are properly ignored.
- moved error parsing/handling to the `close` handler, which also eliminates the potential for console output that too closely matches an "expired" error being interpreted as an error and breaking the session immediately.

## Target release date 🗓️
TBD.

## Preview 📷
| Before  | After   |
| ------- | ------- |
| <img width="215" alt="M3-8722-before" src="https://github.com/user-attachments/assets/3878dfc0-b63b-48b4-a3bd-a00868386f02"> | <img width="323" alt="M3-8722-error" src="https://github.com/user-attachments/assets/69ec9196-29ba-4c2a-8ce7-ac3a17094fb0"> |

## How to test 🧪

### Prerequisites
Running linode(s).

### Reproduction steps
Launch a lish session for a machine, type `^a d kill <enter>`. The session will exit and show a spinning circle.

### Verification steps
- Launch a lish session for a machine, type `^a d kill <enter>`. The session will reconnect.
- Repeat, it should reconnect again.
- Repeat, it should reconnect again.
- Repeat, it should show an error message (if all three attempts are withing 60s).
- Click the "Retry Connection" button. It should reconnect.
- Ensure that the reconnected session actually works, i.e., that interaction with the terminal is successful.

## As an Author I have considered 🤔

*Check all that apply*
